odd-platform-api/src/main/java/org/opendatadiscovery/oddplatform/service/DataSourceServiceImpl.java

```diff
@@ -11,60 +12,135 @@
+                        .flatMap(dsDto -> Mono.zip(
+                            searchEntrypointRepository.updateDataSourceVector(dsDto.dataSource().getId()),
+                            searchEntrypointRepository.updateNamespaceVector(dsDto.namespace().getId())
+                        ).map(t -> dsDto));
+                }
+
+                return update(dataSource.dataSource(), form, null)
+                    .flatMap(dsDto -> Mono.zip(
+                        searchEntrypointRepository.updateDataSourceVector(dsDto.dataSource().getId()),
+                        searchEntrypointRepository.clearNamespaceVector(dsDto.dataSource().getId())
+                    ).map(t -> dsDto));
+            })
+            .map(dataSourceMapper::mapDto);
+    }
+
+    @Override
+    @ReactiveTransactional
+    public Mono<Long> delete(final long id) {
+        return dataEntityRepository.existsByDataSourceId(id)
+            .flatMap(exists -> {
+                if (!exists) {
+                    return dataSourceRepository.delete(id).map(DataSourcePojo::getId);
+                }
+
+                return Mono.error(new IllegalStateException(
+                    "Data source with ID %d cannot be deleted: there are still data sources attached".formatted(id)));
+            });
     }
 
     @Override
-    public Mono<DataSource> regenerateDataSourceToken(final Long dataSourceId) {
-        return Mono.fromCallable(() -> entityRepository.get(dataSourceId))
-            .filter(Optional::isPresent)
-            .switchIfEmpty(Mono.error(new NotFoundException()))
-            .map(Optional::get)
-            .flatMap(dto -> tokenService.regenerateToken(dto.token().tokenPojo()).map(tokenPojo -> {
-                final TokenDto updatedTokenDto = tokenRepository.updateToken(tokenPojo);
-                return entityMapper.mapPojo(new DataSourceDto(dto.dataSource(), dto.namespace(), updatedTokenDto));
-            }));
+    public Mono<DataSource> regenerateDataSourceToken(final long id) {
+        return dataSourceRepository.getDto(id)
+            .switchIfEmpty(Mono.error(new NotFoundException("Data source with id % doesn't exist", id)))
+            .flatMap(dto -> tokenGenerator.regenerateToken(dto.token().tokenPojo())
+                .flatMap(tokenRepository::updateToken)
+                .map(t -> new DataSourceDto(dto.dataSource(), dto.namespace(), t)))
+            .map(dataSourceMapper::mapDto);
+    }
+
+    private Mono<DataSourceDto> update(final DataSourcePojo dataSource,
+                                       final DataSourceUpdateFormData form,
+                                       final NamespacePojo namespace) {
+        return dataSourceRepository.update(dataSourceMapper.applyToPojo(dataSource, form, namespace))
+            .switchIfEmpty(Mono.error(new NotFoundException("Data source with id % doesn't exist", dataSource.getId())))
+            .zipWith(tokenRepository.getByDataSourceId(dataSource.getId()))
```

Comment:
Why are you making another call to tokenRepository if you already got dataSourceDto above and can just pass dto here instead of pojo?